File: viewutils/src/main/java/io/vithor/yamvpframework/viewutils/InputMasker.java
package io.vithor.yamvpframework.viewutils;

import android.text.Editable;
import android.text.TextWatcher;
import android.widget.EditText;

public class InputMasker implements TextWatcher {

    private final String replaceRegex;
    protected boolean isUpdating;
    protected String mOldString = "";
    protected String mMask;
    protected final EditText mEditText;

    public InputMasker(EditText editText, @InputMasks.Mask String mask, String replaceRegex) {
        mEditText = editText;
        mEditText.addTextChangedListener(this);
        mMask = mask;
        this.replaceRegex = replaceRegex;
    }

    @Override
    public void onTextChanged(CharSequence s, int start, int before, int count) {
        String str = s.toString().replaceAll(replaceRegex, "");
        StringBuilder mask = new StringBuilder();
        if (isUpdating) {
            mOldString = str;
            isUpdating = false;
            return;
        }
        int i = 0;
        for (char m : mMask.toCharArray()) {
            if (m != '#' && str.length() > mOldString.length()) {
                mask.append(m);
                continue;
            }
            try {
                mask.append(str.charAt(i));
            } catch (Exception e) {
                break;
            }
            i++;
        }
        isUpdating = true;
        mEditText.setText(mask.toString());
        mEditText.setSelection(mask.length());
    }

    @Override
    public void beforeTextChanged(CharSequence s, int start, int count, int after) {
    }

    @Override
    public void afterTextChanged(Editable s) {
    }

}

File: ui/src/main/java/io/vithor/yamvpframework/ui/InputMasker.kt
package io.vithor.yamvpframework.ui

import android.text.Editable
import android.text.TextWatcher
import android.widget.EditText

open class InputMasker(protected val mEditText: EditText, @InputMasks.Mask protected var mMask: String, private val replaceRegex: String) : TextWatcher {
    protected var isUpdating: Boolean = false
    protected var mOldString = ""

    init {
        mEditText.addTextChangedListener(this)
    }

    override fun onTextChanged(s: CharSequence, start: Int, before: Int, count: Int) {
        val str = s.toString().replace(replaceRegex.toRegex(), "")
        val mask = StringBuilder()
        if (isUpdating) {
            mOldString = str
            isUpdating = false
            return
        }
        var i = 0
        for (m in mMask.toCharArray()) {
            if (m != '#' && str.length > mOldString.length) {
                mask.append(m)
                continue
            }
            try {
                mask.append(str[i])
            } catch (e: Exception) {
                break
            }

            i++
        }
        isUpdating = true
        mEditText.setText(mask.toString())
        mEditText.setSelection(mask.length)
    }

    override fun beforeTextChanged(s: CharSequence, start: Int, count: Int, after: Int) {
    }

    override fun afterTextChanged(s: Editable) {
    }

}
